Do not reload page after adding comments in Pull Request reviews

jpraet opened the pull request:
Fixes #8861.

Use ajax on PR review page for submitting review comments, adding replies and (un)resolving conversations, instead of refreshing the entire page on each interaction.

a1012112796 commented on models/issue_comment.go:
```diff
@@ -1148,6 +1148,66 @@ func fetchCodeCommentsByReview(e Engine, issue *Issue, currentUser *User, review
+		TreePath: treePath,
+		Line:     line,
+	}
+	var comments []*Comment
```
make ?

jpraet replied:
Thank you for the review. I will apply your suggestions. I don't understand what you mean by this comment though?

lunny replied:
Since we can't predict how many comments will be fetched from database, we can't give a meaningful capability on the make.

lafriks replied:
I think he meant that comments variable needs to be initialized using make

zeripath replied:
I think this is fine - AFAIU xorm will generate the comments array as appropriate in Find.

a1012112796 commented on routers/repo/pull_review.go:
```diff
@@ -58,11 +87,32 @@ func CreateCodeComment(ctx *context.Context, form auth.CodeCommentForm) {
+			return
+		}
+		ctx.Data["AfterCommitID"] = pullHeadCommitID
+		ctx.HTML(200, tplConversation)
+	} else {
+		ctx.Redirect(comment.HTMLURL())
+	}
```
```suggestion
		ctx.HTML(200, tplConversation)
		return
	}
	ctx.Redirect(comment.HTMLURL())
```

a1012112796 commented on routers/repo/pull_review.go:
```diff
@@ -103,9 +153,32 @@ func UpdateResolveConversation(ctx *context.Context) {
+			return
+		}
+		ctx.Data["AfterCommitID"] = pullHeadCommitID
+		ctx.HTML(200, tplConversation)
+	} else {
+		ctx.JSON(200, map[string]interface{}{
+			"ok": true,
+		})
+	}
```
```suggestion
		ctx.HTML(200, tplConversation)
		return
	}
	ctx.JSON(200, map[string]interface{}{
		"ok": true,
	})
```

a1012112796 commented on routers/routes/macaron.go:
```diff
@@ -854,6 +854,7 @@ func RegisterMacaronRoutes(m *macaron.Macaron) {
 			m.Group("/files", func() {
 				m.Get("", context.RepoRef(), repo.SetEditorconfigIfExists, repo.SetDiffViewStyle, repo.SetWhitespaceBehavior, repo.ViewPullFiles)
 				m.Group("/reviews", func() {
+					m.Post("/new_comment", repo.RenderNewCodeCommentForm)
```
why not ``Get()`` ?

silverwind commented on routers/repo/pull_review.go:
```diff
@@ -103,6 +153,29 @@ func UpdateResolveConversation(ctx *context.Context) {
+		ctx.Data["CanMarkConversation"] = true
+		ctx.Data["Issue"] = comment.Issue
+		if err = comment.Issue.LoadPullRequest(); err != nil {
+			ctx.ServerError("comment.Issue.LoadPullReqiest", err)
```
```suggestion
			ctx.ServerError("comment.Issue.LoadPullRequest", err)
```

silverwind commented on routers/repo/pull_review.go:
```diff
@@ -58,11 +87,32 @@ func CreateCodeComment(ctx *context.Context, form auth.CodeCommentForm) {
+		ctx.Data["AfterCommitID"] = pullHeadCommitID
+		ctx.HTML(200, tplConversation)
+		return
+	}
```
This block is mostly duplicate with the block below. Maybe it's worth factoring out the duplicated parts into a separate function?

silverwind commented:
Some suggestions. Mostly async ajax requests and simpliciation:

````diff
diff --git a/web_src/js/index.js b/web_src/js/index.js
index fbbd97bdf..b94bf527c 100644
--- a/web_src/js/index.js
+++ b/web_src/js/index.js
@@ -1275,9 +1275,9 @@ function initPullRequestReview() {
     e.preventDefault();
     $(this).closest('.menu').toggle('visible');
   });

-  $('a.add-code-comment').on('click', function (e) {
+  $('a.add-code-comment').on('click', async function (e) {
     if ($(e.target).hasClass('btn-add-single')) return; // https://github.com/go-gitea/gitea/issues/4745
     e.preventDefault();

     const isSplit = $(this).closest('.code-diff').hasClass('code-diff-split');
@@ -1308,22 +1308,20 @@ function initPullRequestReview() {

     const td = ntr.find(`.add-comment-${side}`);
     let commentCloud = td.find('.comment-code-cloud');
     if (commentCloud.length === 0 && !ntr.find('button[name="is_review"]').length) {
-      const url = $(this).data('new-comment-url');
-      $.get(url, (data) => {
-        td.html(data);
-        commentCloud = td.find('.comment-code-cloud');
-        assingMenuAttributes(commentCloud.find('.menu'));
-        td.find("input[name='line']").val(idx);
-        td.find("input[name='side']").val(side === 'left' ? 'previous' : 'proposed');
-        td.find("input[name='path']").val(path);
-        const $textarea = commentCloud.find('textarea');
-        attachTribute($textarea.get(), {mentions: true, emoji: true});
-        const $simplemde = setCommentSimpleMDE($textarea);
-        $textarea.focus();
-        $simplemde.codemirror.focus();
-      });
+      const data = await $.get($(this).data('new-comment-url'));
+      td.html(data);
+      commentCloud = td.find('.comment-code-cloud');
+      assingMenuAttributes(commentCloud.find('.menu'));
+      td.find("input[name='line']").val(idx);
+      td.find("input[name='side']").val(side === 'left' ? 'previous' : 'proposed');
+      td.find("input[name='path']").val(path);
+      const $textarea = commentCloud.find('textarea');
+      attachTribute($textarea.get(), {mentions: true, emoji: true});
+      const $simplemde = setCommentSimpleMDE($textarea);
+      $textarea.focus();
+      $simplemde.codemirror.focus();
     }
   });
 }

@@ -2483,31 +2481,26 @@ $(document).ready(async () => {
     e.checked = false;
     $(e).trigger('click');
   });

-  $(document).on('click', '.resolve-conversation', function (e) {
+  $(document).on('click', '.resolve-conversation', async function (e) {
     e.preventDefault();
-    const id = $(this).data('comment-id');
+    const comment_id = $(this).data('comment-id');
     const origin = $(this).data('origin');
     const action = $(this).data('action');
     const url = $(this).data('update-url');

-    $.post(url, {
-      _csrf: csrf,
-      origin,
-      action,
-      comment_id: id,
-    }, (data) => {
-      if ($(this).closest('.conversation-holder').length) {
-        const conversation = $(data);
-        $(this).closest('.conversation-holder').replaceWith(conversation);
-        conversation.find('.dropdown').dropdown();
-        initReactionSelector(conversation);
-        initClipboard();
-      } else {
-        reload();
-      }
-    });
+    const data = await $.post(url, {_csrf: csrf, origin, action, comment_id});

+    if ($(this).closest('.conversation-holder').length) {
+      const conversation = $(data);
+      $(this).closest('.conversation-holder').replaceWith(conversation);
+      conversation.find('.dropdown').dropdown();
+      initReactionSelector(conversation);
+      initClipboard();
+    } else {
+      reload();
+    }
   });

   buttonsClickOnEnter();
   searchUsers();
@@ -3617,27 +3610,23 @@ function initIssueList() {
 $(document).on('click', 'button[name="is_review"]', (e) => {
   $(e.target).closest('form').append('<input type="hidden" name="is_review" value="true">');
 });

-$(document).on('submit', '.conversation-holder form', (e) => {
+$(document).on('submit', '.conversation-holder form', async (e) => {
   e.preventDefault();
   const form = $(e.target);
-  $.post(form.attr('action'), form.serialize(), (data) => {
-    const conversation = $(data);
-    const path = conversation.data('path');
-    const side = conversation.data('side');
-    const idx = conversation.data('idx');
-    const lineType = form.closest('tr').data('line-type');
-    form.closest('.conversation-holder').replaceWith(conversation);
-    if (lineType === 'same') {
-      $(`a.add-code-comment[data-path="${path}"][data-idx="${idx}"]`).addClass('invisible');
-    } else {
-      $(`a.add-code-comment[data-path="${path}"][data-side="${side}"][data-idx="${idx}"]`).addClass('invisible');
-    }
-    conversation.find('.dropdown').dropdown();
-    initReactionSelector(conversation);
-    initClipboard();
-  });
+  const newConversationHolder = $(await $.post(form.attr('action'), form.serialize()));
+  const {path, side, idx} = newConversationHolder.data();

+  form.closest('.conversation-holder').replaceWith(newConversationHolder);
+  if (form.closest('tr').data('line-type') === 'same') {
+    $(`a.add-code-comment[data-path="${path}"][data-idx="${idx}"]`).addClass('invisible');
+  } else {
+    $(`a.add-code-comment[data-path="${path}"][data-side="${side}"][data-idx="${idx}"]`).addClass('invisible');
+  }
+  newConversationHolder.find('.dropdown').dropdown();
+  initReactionSelector(newConversationHolder);
+  initClipboard();
 });

 window.cancelCodeComment = function (btn) {
   const form = $(btn).closest('form');
````

jpraet commented:
Rebased because there were some merge conflicts. I also extracted some duplicate code. FetchCodeCommentsByLine was initially more or less copied from fetchCodeCommentsByReview. Now they both use a common findCodeComments function instead.

jpraet commented on models/issue_comment.go:
```diff
@@ -1118,7 +1135,19 @@ func fetchCodeCommentsByReview(e Engine, issue *Issue, currentUser *User, review
+			}
+			comment.Review = re
+		}
+		comments[n] = comment
+		n++
```
I'm new to golang. Is this an OK way to remove the pending reviews from the array? I got it from https://github.com/golang/go/wiki/SliceTricks#filter-in-place.

zeripath replied:
This is fine unless the slice is very big initially, much smaller afterwards and likely to hang around for a long time, or very big with only a few removals

* if you think the slice is going to hang around a long time and is quite big initially and small afterwards you should allocate a new a slice of the correct size at the end and issue `copy(filtered, slice)`

* If the slice is very big and is going to have only a few removals you should do: 
```go
	i := 0
	for i < len(slice) {
 		if !removing {
			i++
			continue
		}
		copy(slice[i:], slice[i+1:])
		slice= slice[:len(slice)-1]   
 	}
```

if the slice is expected to be less than 10, as it is,  then this is fine

jpraet commented:
ping 👀

lafriks commented:
🚀
